### PR TITLE
Improve auto-login after self registration if the callback URL in the self-registration confirmation URL contains sessionDataKey

### PR DESCRIPTION
**Pull request by piraveena**
Resolves https://github.com/wso2/product-is/issues/13254

If the **autologin feature** is enabled, 
User will be auto logged in to the app's accessURL if the **accessURL** is configured as sp property. The self-registration confirmation URL will looks line as shown below.
`https://localhost:9443/accountrecoveryendpoint/confirmregistration.do?confirmation=d068262b-7719-4e56-8352-43ac6eff75c9&userstoredomain=PRIMARY&username=userSelf67&tenantdomain=carbon.super&callback=https%3A%2F%2Flocalhost%3A9443%2Fmyaccount%2F`

### Purpose
But if the app is not configured with the accessURL, the user will not be auto-logged in to the app home page. So this PR improves that logic. After merging this PR, if the callbackURL contains sessionDataKey, the user will be auto logged in to the app.

**A sample self-registration confirmation URL is given below:**
` https://localhost:9443/accountrecoveryendpoint/confirmregistration.do?confirmation=06e0b578-86e0-4d1b-b1b9-b02b0b64e2ef&userstoredomain=PRIMARY&username=userself3&tenantdomain=carbon.super&callback=https%3A%2F%2Flocalhost%3A9443%2Fauthenticationendpoint%2Flogin.do%3Fclient_id%3DMY_ACCOUNT%26code_challenge%3DKjiNypeKquWErDvW5OWF-HK22Rhy54T5Bh4VZy9id-4%26code_challenge_method%3DS256%26commonAuthCallerPath%3D%2Foauth2%2Fauthorize%26forceAuth%3Dfalse%26passiveAuth%3Dfalse%26redirect_uri%3Dhttps%3A%2F%2Flocalhost%3A9443%2Fmyaccount%26response_mode%3Dform_post%26response_type%3Dcode%26scope%3Dopenid+openid+SYSTEM%26tenantDomain%3Dcarbon.super%26sessionDataKey%3Dcc3a3b7a-342a-433f-9a28-13ae2695d5ed%26relyingParty%3DMY_ACCOUNT%26type%3Doidc%26sp%3DMy+Account%26isSaaSApp%3Dtrue%26authenticators%3DBasicAuthenticator%3ALOCAL `

### Related Issues
- Issue `#1` or (None)

### Checklist
- [ ] e2e cypress tests locally verified.
- [ ] Manual test round performed and verified.
- [ ] UX/UI review done on final implementation.
- [ ] Documentation provided. (Add links if there's any)
- [ ] Unit tests provided. (Add links if there's any)
- [ ] Integration tests provided. (Add links if there's any)

### Related PRs
- Related PR `#1` or (None)

### Security checks
- [ ] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines?
- [ ] Ran FindSecurityBugs plugin and verified report?
- [ ] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?

### CODE REVIEW

File: apps/recovery-portal/src/main/webapp/self-registration-complete.jsp [hunk 1]
```diff
@@ -22,11 +22,24 @@
...
 <jsp:directive.include file="tenant-resolve.jsp"/>
 <%
     boolean isEmailNotificationEnabled = false;
     String callback = (String) request.getAttribute("callback");
+    String username = request.getParameter("username");
+    String userStoreDomain = request.getParameter("userstoredomain");
+    String sessionDataKey = null;
+    String fullyQualifiedUsername = username;
+
+    boolean isAutoLoginEnable = Boolean.parseBoolean(Utils.getConnectorConfig("SelfRegistration.AutoLogin.Enable",
```
Comment (senthalan) [inline apps/recovery-portal/src/main/webapp/self-registration-complete.jsp:39]: do we need to check this ? if the auto login is disabled, the cookie won't be created in the previous page right ?

Comment (piraveena) [reply apps/recovery-portal/src/main/webapp/self-registration-complete.jsp:39]: yes. It is a redundant one. I will fix it

File: apps/recovery-portal/src/main/webapp/password-reset-complete.jsp [hunk 2]
```diff
@@ -178,6 +191,23 @@
 </head>
 <body>
 
+<div>
+    <form id="callbackForm" name="callbackForm" method="post" action="/commonauth">
+        <%
+            if (username != null) {
```
Comment (chamathns) [inline apps/recovery-portal/src/main/webapp/password-reset-complete.jsp:197]:
```suggestion
            if (StringUtils.isNotBlank(username)) {
```

File: apps/recovery-portal/src/main/webapp/password-reset-complete.jsp [hunk 3]
```diff
@@ -62,6 +62,7 @@
     String username = null;
     PreferenceRetrievalClient preferenceRetrievalClient = new PreferenceRetrievalClient();
     Boolean isAutoLoginEnable = preferenceRetrievalClient.checkAutoLoginAfterPasswordRecoveryEnabled(tenantDomain);
+    String sessionDataKey = null;
```
Comment (chamathns) [inline apps/recovery-portal/src/main/webapp/password-reset-complete.jsp:65]: shall we use `String sessionDataKey = StringUtils.EMPTY;`
or
`String sessionDataKey = "";` instead?

File: apps/recovery-portal/src/main/webapp/self-registration-complete.jsp [hunk 4]
```diff
@@ -36,6 +46,26 @@
 
     isEmailNotificationEnabled = Boolean.parseBoolean(application.getInitParameter(
             IdentityManagementEndpointConstants.ConfigConstants.ENABLE_EMAIL_NOTIFICATION));
+    boolean isSessionDataKeyPresent = false;
+    if (userStoreDomain != null) {
```
Comment (chamathns) [inline apps/recovery-portal/src/main/webapp/self-registration-complete.jsp:50]:
```suggestion
    if (StringUtils.isNotBlank(userStoreDomain)) {
```

File: apps/recovery-portal/src/main/webapp/self-registration-complete.jsp [hunk 5]
```diff
@@ -22,11 +22,21 @@
...
+<%@ page import="java.util.HashMap" %>
+<%@ page import="java.util.Map" %>
 <jsp:directive.include file="includes/localize.jsp"/>
 <jsp:directive.include file="tenant-resolve.jsp"/>
 <%
     boolean isEmailNotificationEnabled = false;
     String callback = (String) request.getAttribute("callback");
+    String username = request.getParameter("username");
+    String userStoreDomain = request.getParameter("userstoredomain");
+    String sessionDataKey = null;
```
Comment (chamathns) [inline apps/recovery-portal/src/main/webapp/self-registration-complete.jsp:36]: same as above

File: apps/recovery-portal/src/main/webapp/self-registration-complete.jsp [hunk 6]
```diff
@@ -36,6 +46,26 @@
 
     isEmailNotificationEnabled = Boolean.parseBoolean(application.getInitParameter(
             IdentityManagementEndpointConstants.ConfigConstants.ENABLE_EMAIL_NOTIFICATION));
+    boolean isSessionDataKeyPresent = false;
```
Comment (chamathns) [inline apps/recovery-portal/src/main/webapp/self-registration-complete.jsp:49]: false is redundant

File: apps/recovery-portal/src/main/webapp/password-reset-complete.jsp [hunk 7]
```diff
@@ -62,6 +62,7 @@
     String username = null;
     PreferenceRetrievalClient preferenceRetrievalClient = new PreferenceRetrievalClient();
     Boolean isAutoLoginEnable = preferenceRetrievalClient.checkAutoLoginAfterPasswordRecoveryEnabled(tenantDomain);
+    String sessionDataKey = StringUtils.EMPTY;;
```
Comment (chamathns) [inline apps/recovery-portal/src/main/webapp/password-reset-complete.jsp:65]:
```suggestion
    String sessionDataKey = StringUtils.EMPTY;
```

**Comment by piraveena** [comment]
PR builder was successfulr https://jenkins-support-all.wso2.com/jenkins/job/DEV_PR_BUILDER/1540/console

**Review by chamathns — APPROVED** [review]

**Pull request merged**